File: Array & Hashing/287. Find the Duplicate Number/find_the_duplicate_number.py
class Solution:
    def findDuplicate(self, nums: List[int], start: int = 0) -> int:
      def swap(arr: List[int], a: int, b: int) -> None:
        arr[a], arr[b] = arr[b], arr[a]
      
      def is_same() -> bool:
        return nums[start] == nums[nums[start]]

      while not is_same():
        swap(nums, start, nums[start])
      
      return nums[start]
      

File: Array & Hashing/287. Find the Duplicate Number/findTheDuplicateNumber.ts
const findDuplicate = function(nums: number[], start: number = 0): number {
  const swap = (arr: number[], a: number, b: number): [number, number] => [arr[a], arr[b]] = [arr[b], arr[a]];

  const isSame = (): boolean => nums[start] === nums[nums[start]];
  while (!isSame()) {
    swap(nums, start, nums[start]);
  } 

  return nums[start];
};